### PR TITLE
Fix #24707 url resolution: better error for uncallable view passed via import

### PR DESCRIPTION
**Pull request by hjwp**

### CODE REVIEW

File: django/core/urlresolvers.py [hunk 1]
```diff
@@ -94,6 +95,11 @@ def get_callable(lookup_view, can_fail=False):
     if callable(lookup_view):
         return lookup_view
 
+    if not isinstance(lookup_view, string_types):
```
Comment (knbk) [inline django/core/urlresolvers.py]: Use `six.string_types` from `django.utils.six` (which is already imported).

File: django/core/urlresolvers.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 import re
 import warnings
 from importlib import import_module
+from six import string_types
```
Comment (knbk) [inline django/core/urlresolvers.py]: As below, use `django.utils.six`.

**Comment by hjwp** [comment]
there you go!  Or would you rather have me squash it into a single commit?

**Comment by MarkusH** [comment]
Please squash the commits into a single commit and use past tense in the commit message.

**Comment by MarkusH** [comment]
buildbot, ok to test.

**Comment by hjwp** [comment]
how 'bout that? :)

File: django/core/urlresolvers.py [hunk 3]
```diff
@@ -94,6 +94,11 @@ def get_callable(lookup_view, can_fail=False):
     if callable(lookup_view):
         return lookup_view
 
+    if not isinstance(lookup_view, six.string_types):
+        raise ViewDoesNotExist(
+            "'%s' was not a callable or a dot-notation path" % lookup_view
```
Comment (MarkusH) [inline django/core/urlresolvers.py]: I'd use present tense here: "... is not a callable ..."

File: tests/urlpatterns_reverse/erroneous_urls.py [hunk 4]
```diff
@@ -19,8 +19,10 @@
         url(r'erroneous_unqualified/$', 'unqualified_view'),
         # View does not exist
         url(r'missing_inner/$', 'urlpatterns_reverse.views.missing_view'),
-        # View is not callable
-        url(r'uncallable/$', 'urlpatterns_reverse.views.uncallable'),
+        # View is not callable (dot-notation)
+        url(r'uncallable-dotted/$', 'urlpatterns_reverse.views.uncallable'),
+        # View is not callable (view object)
```
Comment (MarkusH) [inline tests/urlpatterns_reverse/erroneous_urls.py]: How about `# View is not a callable (explicit import; arbitrary Python object)`

File: tests/urlpatterns_reverse/erroneous_urls.py [hunk 5]
```diff
@@ -19,8 +19,10 @@
         url(r'erroneous_unqualified/$', 'unqualified_view'),
         # View does not exist
         url(r'missing_inner/$', 'urlpatterns_reverse.views.missing_view'),
-        # View is not callable
-        url(r'uncallable/$', 'urlpatterns_reverse.views.uncallable'),
+        # View is not callable (dot-notation)
```
Comment (MarkusH) [inline tests/urlpatterns_reverse/erroneous_urls.py]: How about # View is not a callable (string import; arbitrary Python object)

**Comment by MarkusH** [comment]
I'd change the commit message to something along the lines "Fixed #24707 -- Improved error reporting for explicitly imported uncallable views"

**Comment by hjwp** [comment]
here you go!

**Comment by MarkusH** [comment]
Let's see what Jenkins thinks about the patch. But LGTM

**Comment by hjwp** [comment]
Thanks for taking the time to help Markus.  Much appreciated.

**Comment by MarkusH** [comment]
Merged in 40768ec29dc0a38c3b824b32c1a9167c0e8b5ee0. Thanks :smile:

**Pull request closed without merging**